test/lifecycle_module_test.dart

```diff
@@ -948,20 +909,188 @@ void main() {
+              // those events don't affect the test expectations.
+              // ignore: unawaited_futures
+              module.didLoad.first.then((_) => module.eventList.clear());
+            } else {
+              await gotoState(module, state);
             }
-          }, count: 2));
+            module.eventList.clear();
+          });
+
+          test('should unload and then dispose', () async {
+            expectInLifecycleState(module, state);
+            await module.dispose();
+            expect(module.isDisposed, isTrue);
+            expect(
+                module.eventList,
+                equals([]
+                  ..addAll(expectedPreUnloadStates)
+                  ..addAll(
+                      ['onShouldUnload', 'willUnload', 'onUnload', 'didUnload'])
+                  ..addAll(expectedDisposalStates)));
+          });
+
+          group('with onShouldUnload=false', () {
+            setUp(() {
+              module.mockShouldUnload = false;
+            });
+
+            test('should dispose despite the unload being canceled', () async {
+              expectInLifecycleState(module, state);
+              await module.dispose();
+              expect(module.isDisposed, isTrue);
+              expect(
+                  module.eventList,
+                  equals([]
+                    ..addAll(expectedPreUnloadStates)
+                    ..addAll(['onShouldUnload'])
+                    ..addAll(expectedDisposalStates)));
+            });
+
+            test(
+                'should warn that the unload was canceled but that disposal will continue',
+                () async {
+              expect(
+                  Logger.root.onRecord,
+                  emitsThrough(logRecord(
+                      level: Level.WARNING,
+                      message: contains(
+                          '.dispose() was called but Module "${module.name}" canceled'))));
+              expectInLifecycleState(module, state);
+              await module.dispose();
+            });
+          });
+
+          group('with an onUnload that throws', () {
+            setUp(() {
+              module.onUnloadError = testError;
+            });
+
+            test('should emit the unload failure from didUnload', () async {
+              expect(
+                  module.didUnload.first, throwsA(same(module.onUnloadError)));
+              expectInLifecycleState(module, state);
+              await module.dispose();
+            });
+
+            test('should dispose despite the unload failing', () async {
+              expectInLifecycleState(module, state);
+              await module.dispose();
+              expect(module.isDisposed, isTrue);
+              expect(
+                  module.eventList,
+                  equals([]
+                    ..addAll(expectedPreUnloadStates)
+                    ..addAll(['onShouldUnload', 'willUnload'])
+                    ..addAll(expectedDisposalStates)));
+            });
```

Comment:
Do we have a test somewhere that verifies that this works properly for child modules as well? I think we manage the child modules, so it is probably fine.